### ISSUE TITLE
Race Condition breaks push from subgit folder

### ISSUE THREAD

**Issue by samsieber** [samsieber/subgit-sync#1]
For this explanation, assume that a/b/c is mapped in the upstream to the root of the subgit

When you push commits from a subgit folder, the basic flow is:
1) Check if the `upstream repo` has commits that haven’t been imported
2) If there are commits that haven’t been imported, import them now
3) If any freshly imported commits affected the `subgit repo`, abort (e.g. if any of the imported commits had files in the `a/b/c` folder - because actually generate new commits in the `subgit repo` repo).
4) If we’ve made it this far, the mapping between `upstream repo` and the `subgit repo` is up to date - we apply all the changes to the latest commit in the `upstream repo` in our working copy
5) We push the new commits to the master branch in `upstream repo`.

Basically, if someone pushes to upstream while the update hook is performing step 4, step 5 will fail because it will be pointing to old commits. Even worse, we already permanently stored the mapping for the commits during step 4, so subsequent pushes of that exact subgit commit (through the update hook), because when we rerun the process we find the old (but incorrect mapping) and reuse it.

So the solution is to back out new created mappings of step 4 if step 5 fails.
